engine/src/main/java/org/hibernate/validator/internal/properties/javabean/JavaBeanExecutable.java

```diff
@@ -202,24 +207,53 @@ public String toString() {
 		}
 		else {
 			// in this case, we have synthetic or implicit parameters
+
+			// do we have the information about which parameter is synthetic/implicit?
+			// (this metadata is only included when classes are compiled with the '-parameters' flag)
+			boolean hasParameterModifierInfo = isAnyParameterCarryingMetadata( parameterArray );
+
+			if ( ! hasParameterModifierInfo ) {
+				LOG.missingParameterMetadataWithSyntheticOrImplicitParameters( executable );
+			}
+
 			int explicitlyDeclaredParameterIndex = 0;
 
 			for ( int i = 0; i < parameterArray.length; i++ ) {
-				if ( parameterArray[i].isSynthetic() || parameterArray[i].isImplicit() ) {
-					// in this case, the parameter is not present in genericParameterTypes
-					parameters.add( new JavaBeanParameter( i, parameterArray[i], parameterTypes[i], parameterTypes[i] ) );
-				}
-				else {
+				if ( explicitlyDeclaredParameterIndex < genericParameterTypes.length // we might already be out of the bounds of generic params array
+						&& isExplicit( parameterArray[i] )
+						&& parameterTypesMatch( parameterTypes[i], genericParameterTypes[explicitlyDeclaredParameterIndex] ) ) {
+					// in this case we have a parameter that is present and matches ("most likely") to the one in the generic parameter types list
 					parameters.add( new JavaBeanParameter( i, parameterArray[i], parameterTypes[i],
 							getErasedTypeIfTypeVariable( genericParameterTypes[explicitlyDeclaredParameterIndex] ) ) );
 					explicitlyDeclaredParameterIndex++;
 				}
+				else {
+					// in this case, the parameter is not present in genericParameterTypes, or the types doesn't match
+					parameters.add( new JavaBeanParameter( i, parameterArray[i], parameterTypes[i], parameterTypes[i] ) );
+				}
 			}
 		}
 
 		return CollectionHelper.toImmutableList( parameters );
 	}
 
+	private static boolean isAnyParameterCarryingMetadata(Parameter[] parameterArray) {
+		for ( Parameter parameter : parameterArray ) {
+			if ( parameter.isSynthetic() || parameter.isImplicit() ) {
+				return true;
+			}
+		}
+		return false;
```

Comment:
Hmmm, what if the method is fully static? We won't have any synthetic or implicit parameters or will we?

Reply:
Ah, yes, I missed that from the diff visible on GH, sorry :)